Backfill_Correction/delphiBackfillCorrection/R/model.R

```diff
@@ -0,0 +1,259 @@
+    test_data[paste0("sqrty", as.character(split))] = 0
+    qv_pre = max_raw * split * 0.2
+    qv_next = max_raw * (split+1) * 0.2
+
+    train_data[(train_data[[value_col]] <= (qv_next)^2)
+               & (train_data[[value_col]] > (qv_pre)^2), 
+               paste0("sqrty", as.character(split))] = 1
+    test_data[(test_data[[value_col]] <= (qv_next)^2)
+              & (test_data[[value_col]] > (qv_pre)^2), 
+              paste0("sqrty", as.character(split))] = 1
+    sqrtscale[split+1] = paste0("sqrty", as.character(split))
+  }
+  return (list(train_data, test_data, sqrtscale))
+}
+
+#' Fetch model and use to generate predictions/perform corrections
+#'
+#' @template train_data-template
+#' @param test_data Data frame for testing 
+#' @template taus-template
+#' @template covariates-template
+#' @template lp_solver-template
+#' @template lambda-template
+#' @template geo_level-template
+#' @template geo-template
+#' @template indicator-template
+#' @template signal-template
+#' @template signal_suffix-template
+#' @template value_type-template
+#' @template test_lag-template
+#' @template train_models-template
+#' @template make_predictions-template
+#' @param model_save_dir directory containing trained models
+#' @param training_end_date Most recent training date
+#'
+#' @importFrom stats predict coef
+#'
+#' @export
+model_training_and_testing <- function(train_data, test_data, taus, covariates,
+                                       lp_solver, lambda, test_lag,
+                                       geo, value_type, model_save_dir, 
+                                       indicator, signal, 
+                                       geo_level, signal_suffix,
+                                       training_end_date, 
+                                       train_models = TRUE,
+                                       make_predictions = TRUE) {
+  success = 0
+  coefs_result = list()
+  coef_list = c("intercept", paste(covariates, '_coef', sep=''))
+  for (tau in taus) {
+    tryCatch(
+      expr = {
+        model_file_name <- generate_filename(indicator, signal, 
+                                        geo_level, signal_suffix, lambda,
+                                        training_end_date, geo, 
+                                        value_type, test_lag, tau)
+        model_path <- file.path(model_save_dir, model_file_name)
+        obj <- get_model(model_path, train_data, covariates, tau,
+                         lambda, lp_solver, train_models=TRUE) 
+
+        if (make_predictions) {
+          y_hat_all = as.numeric(predict(obj, newx = as.matrix(test_data[covariates])))
+          test_data[[paste0("predicted_tau", as.character(tau))]] = y_hat_all
+
+          coefs_result[[success+1]] = coef(obj)
+        }
+
+        success = success + 1
+      },
+      error=function(e) {print(paste("Training failed for", model_path, sep=" "))}
+    )
+  }
+  if (success < 9) {return (NULL)}
```

Comment:
I initially design it to include as detailed information as it can (especially for tooling usage). And if a model training is fail, it's super fast, won't take much time. 

But you are right, here we can stop once there is a failure

Reply:
9 is the length of the default `taus` list. Switched this to actually checking the length of the list.